DataWorkerService/Services/Implementations/OsuApiDataParserService.cs

```diff
@@ -263,25 +264,41 @@ public IEnumerable<GameScore> CreateScores(IEnumerable<ApiGameScore> apiScores)
         return scores;
     }
 
-    public static void ParseBeatmapPartial(Beatmap beatmap, ApiBeatmap apiBeatmap)
+    public Task ParseBeatmapPartial(Beatmap beatmap, ApiBeatmap apiBeatmap)
     {
         beatmap.OsuId = apiBeatmap.Id;
         beatmap.Ruleset = apiBeatmap.Ruleset;
-        beatmap.Length = apiBeatmap.TotalLength;
+        beatmap.TotalLength = apiBeatmap.TotalLength;
         beatmap.DiffName = apiBeatmap.DifficultyName;
 
-        if (apiBeatmap.Beatmapset is null)
+        if (apiBeatmap.Beatmapset?.CreatorId is null)
         {
-            return;
+            return Task.CompletedTask;
         }
 
-        beatmap.Artist = apiBeatmap.Beatmapset.Artist;
-        beatmap.Title = apiBeatmap.Beatmapset.Title;
-        beatmap.MapperName = apiBeatmap.Beatmapset.Creator ?? string.Empty;
-        beatmap.MapperId = apiBeatmap.Beatmapset.CreatorId ?? default;
+        Player? creatorPlayer = _playersCache.Find(p => p.OsuId == apiBeatmap.Beatmapset.CreatorId.Value) ?? throw new Exception(
+                $"Failed to find player with osu! id {apiBeatmap.Beatmapset.CreatorId.Value} in players cache!");
+
+
+        // Create or update BeatmapSet
+        var beatmapSet = new BeatmapSet
+        {
+            OsuId = apiBeatmap.Beatmapset.Id,
+            Artist = apiBeatmap.Beatmapset.Artist,
+            Title = apiBeatmap.Beatmapset.Title,
+            CreatorId = creatorPlayer.Id,
+            RankedStatus = apiBeatmap.Beatmapset.RankedStatus,
+            SubmittedDate = apiBeatmap.Beatmapset.SubmittedDate,
+            RankedDate = apiBeatmap.Beatmapset.RankedDate
+        };
+
+        beatmap.BeatmapSet = beatmapSet;
+        beatmap.BeatmapSetId = beatmapSet.Id;
```

Comment:
```suggestion
        beatmap.BeatmapSet = new BeatmapSet
        {
            OsuId = apiBeatmap.Beatmapset.Id,
            Artist = apiBeatmap.Beatmapset.Artist,
            Title = apiBeatmap.Beatmapset.Title,
            CreatorId = creatorPlayer.Id,
            RankedStatus = apiBeatmap.Beatmapset.RankedStatus,
            SubmittedDate = apiBeatmap.Beatmapset.SubmittedDate,
            RankedDate = apiBeatmap.Beatmapset.RankedDate
        };
```
`BeatmapSet.Id` is not available at this point. Better to create the entity by proxy and let EF assign the Id. Your code likely works fine but this is cleaner imo